runscope/runscope.py

```diff
@@ -0,0 +1,162 @@
+          return False
+    return True
+   
+def get_cached_test_meta_info():
+    try :
+        with open(cache_file) as test_mi:
+            if test_mi is None:
+                return None
+            else:
+                mi = json.load(test_mi)
+                test_mi.close()
+                return mi
+    except ValueError as ve:
+        return None
+    except Exception as e:
+        handle_error("error reading from cache: " + str(e))
+
+def handle_error(error_message):
+    sys.stderr.write("ERROR:|Runscope| " + error_message)
+    sys.exit(1)
+    
+def get_runscope_api_response(url):
+    try :
+        headers = {}
+        headers["Authorization"] = "Bearer {} ".format(args.token[0])
+        headers["Content-type"] = "application/json"
+        headers["Accept"] = "application/json"
+        response = requests.get(url,headers=headers)
+        if response.status_code == 200:
+            return json.loads(response.text)
+        else:
+            em = "error connecting to runscope {}".format(response.status_code)
+            handle_error(em)       
+    except Exception as e:
+        handle_error("error while conneecting to runscope: " + str(e))        
+
+def get_test_bucket_details():
+    test_bucket_details = []
+    response = get_runscope_api_response(runscope_service_url)
+    for data in response['data']:
+        test_bucket_meta_info = {}
+        test_bucket_meta_info['bucket_name'] = data['name']
+        test_bucket_meta_info['tests_url'] = data['tests_url']
+        test_bucket_details.append(test_bucket_meta_info)
+    return test_bucket_details
+
+def get_test_details(test_bucket_list):
+    test_details = []
+    for bucket in test_bucket_list:
+        response = get_runscope_api_response(bucket['tests_url'])
+        for test in response['data']:
+            test_meta_info = {}
+            test_meta_info['latest_result_url'] = bucket['tests_url'] + "/{}/results/latest".format(test['last_run']['test_uuid'])
+            test_meta_info['test_name'] = test['name']
+            test_meta_info['test_bucket_name'] = bucket['bucket_name']
+            test_details.append(test_meta_info)
+    return test_details
+
+def check_and_add(dest_dict,dest_k,val):
+    if val is not None:
+       dest_dict[dest_k] = val
+
+def get_test_latest_result(test_details):
+    uber_result = []
+    for test in test_details:
+        response = get_runscope_api_response(test['latest_result_url'])
+        for request in response['data']['requests']:
+            service_metric = {}
+            service_url = request['url']
+            if service_url is not None:
+               #Need to check on the tag length         
+              # request_result = str(request['result'])
+              service_metric['service_url'] = service_url[:240]
+              check_and_add(service_metric,'request_result',request['result'])
+              check_and_add(service_metric,'request_uuid',request['uuid'])
```

Comment:
@srinivas-kandula This will still throw `KeyError`'s. For example, If there is no `uuid` key in the `request` dict etc, this will throw and not process rest of the keys.

You could pass in the input dict and input keys to the check_and_add function, for example:
```
def check_and_add(dest_dict,dest_k, input_dict, first_key, second_key=None):
    val = input_dict.get(first_key, None)
    if val and second_key:
        val = val.get(second_key, None)
    if val:
        dest_dict[dest_k] = val
```

Reply:
@vikramraman All the key value pairs are part of the runscope API contract as given in the link https://www.runscope.com/docs/api/results either they are mandatory for the valid test to be created or system generated fields so i feel they will always be present in the API response.

Reply:
@srinivas-kandula Even if one of them is missing/changes, we will currently throw with no metrics reported. Safer and more robust to account for failures.